src/main/java/io/aiven/kafka/auth/audit/Auditor.java

```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (c) 2020 Aiven, Helsinki, Finland. https://aiven.io/
+ */
+
+package io.aiven.kafka.auth.audit;
+
+import java.io.Serializable;
+import java.net.InetAddress;
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import kafka.network.RequestChannel.Session;
+import kafka.security.auth.Operation;
+import kafka.security.auth.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class Auditor implements Configurable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(Auditor.class);
+
+    public static final String AUDITOR_AGG_PERIOD_CONF = "aiven.auditor.aggregation.period";
+
+    protected Map<AuditKey, UserActivity> auditStorage = new HashMap<>();
+
+    private final Lock auditLock = new ReentrantLock(true);
```

Comment:
Probably this lock should be unfair: fairness comes with a performance price, but doesn't bring any benefit in this case.